inventory_management_api/routes/inventory.py

```diff
@@ -0,0 +1,120 @@
+            """
+            cursor = execute_query(shard, query, (product_id, per_page, offset))
+            inventory = cursor.fetchall()
+
+            if inventory:
+                self.cache.setex(cache_key, 300, str(inventory))  # Cache for 5 minutes
+                return jsonify({"inventory": inventory})
+            
+            return jsonify({"error": "Product not found"}), 404
+        except mysql.connector.Error as err:
+            return jsonify({"error": str(err)}), 500
+
+    def add_inventory(self):
+        try:
+            data = request.json
+            product_id = data.get('product_id')
+            quantity = data.get('quantity')
+            
+            if not product_id or not quantity:
+                return jsonify({"error": "Invalid data"}), 400
+            
+            shard = get_shard(int(product_id), self.shards)
+            query = "INSERT INTO inventory (product_id, quantity) VALUES (%s, %s)"
+            cursor = execute_query(shard, query, (product_id, quantity))
+            shard.commit()
+            
+            return jsonify({"message": "Inventory added successfully"}), 201
+        except mysql.connector.Error as err:
+            return jsonify({"error": str(err)}), 500
+
+    def update_inventory(self):
+        try:
+            data = request.json
+            product_id = data.get('product_id')
+            quantity = data.get('quantity')
+
+            if not product_id or not quantity:
+                return jsonify({"error": "Invalid data"}), 400
+
+            shard = get_shard(int(product_id), self.shards)
+            query = "UPDATE inventory SET quantity = %s WHERE product_id = %s"
+            cursor = execute_query(shard, query, (quantity, product_id))
+            shard.commit()
+
+            # Invalidate cache after update
+            self.cache.delete(f"inventory_{product_id}")
+
+            return jsonify({"message": "Inventory updated successfully"}), 200
+        except mysql.connector.Error as err:
+            return jsonify({"error": str(err)}), 500
+
+    def delete_inventory(self):
+        try:
+            data = request.json
+            product_id = data.get('product_id')
+
+            if not product_id:
+                return jsonify({"error": "Invalid data"}), 400
+
+            shard = get_shard(int(product_id), self.shards)
+            query = "DELETE FROM inventory WHERE product_id = %s"
+            cursor = execute_query(shard, query, (product_id,))
+            shard.commit()
+
+            # Invalidate cache after delete
+            self.cache.delete(f"inventory_{product_id}")
+
+            # Confirm deletion
+            if cursor.rowcount == 0:
+                return jsonify({"error": "Product not found"}), 404
+
+            return jsonify({"message": "Inventory deleted successfully"}), 200
+        except mysql.connector.Error as err:
+            return jsonify({"error": str(err)}), 500
+
+inventory_routes = InventoryRoutes()
```

Comment:
**Code Structure Issue**: The instantiation of the InventoryRoutes class is done at the module level, which can lead to unintended side effects if the module is imported multiple times or in different contexts. <br> **Fix**: Move the instantiation of the InventoryRoutes class into a function or a conditional block that checks if the module is being run as the main program. This ensures that the class is only instantiated when necessary and avoids potential issues with multiple instantiations in different parts of the application. <br> **Code Suggestion**: 
 ```
 - inventory_routes = InventoryRoutes()
 + if __name__ == '__main__':
 +     inventory_routes = InventoryRoutes()
 ```